### ISSUE TITLE
Filter branches by name and label failing with 404

### ISSUE THREAD

**Issue by Zava2012** [jenkinsci/gitlab-plugin#527]
### Context
- **Gitlab plugin version**: 1.4.5
- **Gitlab version**: 8.17.3
- **Jenkins version**: 2.50
- **Job type**: Freestyle

### Logs & Traces

Jenkins System Log:
```
Call GitLab:
HTTP method: GET
URL: https://MyGitlab/api/v3/projects/TestUser%2FTestRepo/labels
Request headers: [
Accept = [application/json],
PRIVATE-TOKEN = [****FILTERED****]
]
Mar 17, 2017 7:07:50 PM FINEST com.dabsquared.gitlabjenkins.gitlab.GitLabClientBuilder
Got response from GitLab:
URL: https://MyGitlab/api/v3/projects/TestUser%2FTestRepo/labels
Status: 404 Not Found
Response headers: [
Cache-Control = [no-cache],
Connection = [keep-alive],
Content-Length = [25],
Content-Type = [application/json],
Date = [Fri, 17 Mar 2017 16:07:49 GMT],
Server = [nginx],
Vary = [Origin],
X-Request-Id = [01d09a60-a869-422f-87f3-8944a6e11b9e],
X-Runtime = [0.003743]
]
Response body: {
  "error" : "404 Not Found"
}
Mar 17, 2017 7:07:50 PM FINEST com.dabsquared.gitlabjenkins.gitlab.GitLabClientBuilder
Call GitLab:
HTTP method: GET
URL: https://MyGitlab/api/v3/projects/TestUser%2FTestRepo/labels
Request headers: [
Accept = [application/json],
PRIVATE-TOKEN = [****FILTERED****]
]
Mar 17, 2017 7:07:50 PM FINEST com.dabsquared.gitlabjenkins.gitlab.GitLabClientBuilder
Got response from GitLab:
URL: https://MyGitlab/api/v3/projects/TestUser%2FTestRepo/labels
Status: 404 Not Found
Response headers: [
Cache-Control = [no-cache],
Connection = [keep-alive],
Content-Length = [25],
Content-Type = [application/json],
Date = [Fri, 17 Mar 2017 16:07:50 GMT],
Server = [nginx],
Vary = [Origin],
X-Request-Id = [5f588972-3c0d-4aca-8c2b-492c134261f8],
X-Runtime = [0.003752]
]
Response body: {
  "error" : "404 Not Found"
}
Mar 17, 2017 7:07:50 PM FINEST com.dabsquared.gitlabjenkins.gitlab.GitLabClientBuilder
Call GitLab:
HTTP method: GET
URL: https://MyGitlab/api/v3/projects/TestUser%2FTestRepo/labels
Request headers: [
Accept = [application/json],
PRIVATE-TOKEN = [****FILTERED****]
]
Mar 17, 2017 7:07:50 PM FINEST com.dabsquared.gitlabjenkins.gitlab.GitLabClientBuilder
Got response from GitLab:
URL: https://MyGitlab/api/v3/projects/TestUser%2FTestRepo/labels
Status: 404 Not Found
Response headers: [
Cache-Control = [no-cache],
Connection = [keep-alive],
Content-Length = [25],
Content-Type = [application/json],
Date = [Fri, 17 Mar 2017 16:07:50 GMT],
Server = [nginx],
Vary = [Origin],
X-Request-Id = [fc554e53-8b37-49c8-98b9-4dda98bdd09c],
X-Runtime = [0.003708]
]
Response body: {
  "error" : "404 Not Found"
}
Mar 17, 2017 7:07:50 PM FINEST com.dabsquared.gitlabjenkins.gitlab.GitLabClientBuilder
Call GitLab:
HTTP method: GET
URL: https://MyGitlab/api/v3/projects/TestUser%2FTestRepo/labels
Request headers: [
Accept = [application/json],
PRIVATE-TOKEN = [****FILTERED****]
]
Mar 17, 2017 7:07:51 PM FINEST com.dabsquared.gitlabjenkins.gitlab.GitLabClientBuilder
Got response from GitLab:
URL: https://MyGitlab/api/v3/projects/TestUser%2FTestRepo/labels
Status: 404 Not Found
Response headers: [
Cache-Control = [no-cache],
Connection = [keep-alive],
Content-Length = [25],
Content-Type = [application/json],
Date = [Fri, 17 Mar 2017 16:07:50 GMT],
Server = [nginx],
Vary = [Origin],
X-Request-Id = [b1db7700-99b0-4f04-b461-1eba7ee8544e],
X-Runtime = [0.003856]
]
Response body: {
  "error" : "404 Not Found"
}
```

### Problem description

Good evening everyone!
I apologize in advance for my bad English and the completion of this issue, because this is my first issue for the whole time.

The bug is that the plugin does not work **Filter branches by name** and **Filter merge request by label**. The reason for this, based on logs, is that the plugin refers to the wrong URL path. Logs for **Filter branches by name** and **Filter merge request by label** differ only in URL path, if not mistaken. The path URL pattern is shown below:

For **Filter merge request by name**: https://MyGitlab/api/v3/projects/UserName%2FRepoName/repository/branches
For **Filter merge request by label**: https://MyGitlab/api/v3/projects/UserName%2FRepoName/labels 

For some reason, unicode **%2F** instead of a normal slash character, but that's not the point.
If the place of the part above the written templates **UserName%2FRepoName** is replaced with **ProjectID** everything will be fine. That is, the URL path will be found and, for sure, the labels too. The correct template that should be:

For **Filter merge request by name**: https://MyGitlab/api/v3/projects/ProjectID/repository/branches
For **Filter merge request by label**: https://MyGitlab/api/v3/projects/ProjectID/labels 

I'm not sure if this is the problem in the plugin, but there is such a possibility.
If the problem with this and in the code fixes this fairly quickly, then I think that this fix will be useful to many.

Thank you for attention!

**Comment by josemalonsom** [comment]
Same problem here.

### Context
- **Gitlab plugin version**: 1.4.5
- **Gitlab version**: 8.15.8
- **Jenkins version**: 2.32.3
- **Job type**: Pipeline

**Comment by omehegan** [comment]
This issue is usually caused by an Apache reverse proxy in front of your GitLab install. See comments in #399.

**Comment by Zava2012** [comment]
@omehegan, thank you for the answer, but it's not an apache reverse proxy, i checked it.
And thanks for the update!
Now I use the following:

### Context
- **Gitlab plugin version**: 1.4.6
- **Gitlab version**: 9.3.5 CE
- **Jenkins version**: 2.60.1
- **Job type**: Freestyle

The new plugin update has fixed this issue and now **Filter branches by name** and **Filter merge request by label** define and offer to select branches name and labels respectively, but the plugin has one problem left: the trigger does not start the job if the checkbox on option **Filter merge request by label** is set. Without this option everything works fine!

**Comment by omehegan** [comment]
I'm unsure why the plugin upgrade would have fixed this. Can you capture another log for the broken "Filter merge request by label" feature? Also note there is this bug in 1.4.6 which we are testing out a fix for: #583

**Comment by Zava2012** [comment]
However, something was fixed that solved the initially described problem with "404 Not Found".
I use only the function **Opened Merge Request Events** and the build is successful for this event, but after updating the plugin, all jobs (in these jobs i'm using the **Gitlab Merge Requests Builder** plugin and use it's variables) started build with failure NPE (NullPointerException) error, which is shown in problem #583. 

Here is a full log with the **Filter merge request by label** option enabled.
As can be seen from the log, there are no errors or anything like that, but the job does not start the build. All the paths in the logs are correct, I checked it twice.
Updating the plugin to snapshot 1.4.7 did not solve the problem.

```
Jul 13, 2017 3:02:37 AM INFO com.dabsquared.gitlabjenkins.webhook.GitLabWebHook getDynamic
WebHook called with url: /project/MyProject/TEST
Jul 13, 2017 3:02:37 AM FINE com.dabsquared.gitlabjenkins.webhook.build.MergeRequestBuildAction
MergeRequest: {
  "object_kind" : "merge_request",
  "user" : {
    "name" : "MyName",
    "username" : "MyUserName",
    "avatar_url" : "/uploads/system/user/avatar/31/MyUserName.jpg"
  },
  "project" : {
    "name" : "GitLabPluginTest",
    "description" : "",
    "web_url" : "https://MyGitLab/MyUserName/GitLabPluginTest",
    "avatar_url" : null,
    "git_ssh_url" : "git@MyGitLab/MyUserName/GitLabPluginTest.git",
    "git_http_url" : "https://MyGitLab/MyUserName/GitLabPluginTest.git",
    "namespace" : "MyUserName",
    "visibility_level" : 0,
    "path_with_namespace" : "MyUserName/GitLabPluginTest",
    "default_branch" : "development",
    "homepage" : "https://MyGitLab/MyUserName/GitLabPluginTest",
    "url" : "git@MyGitLab/MyUserName/GitLabPluginTest.git",
    "ssh_url" : "git@MyGitLab/MyUserName/GitLabPluginTest.git",
    "http_url" : "https://MyGitLab/MyUserName/GitLabPluginTest.git"
  },
  "object_attributes" : {
    "id" : 1052,
    "target_branch" : "release",
    "source_branch" : "development",
    "source_project_id" : 143,
    "author_id" : 31,
    "assignee_id" : null,
    "title" : "Development",
    "created_at" : "2017-07-13 03:02:37 +0300",
    "updated_at" : "2017-07-13 03:02:37 +0300",
    "milestone_id" : null,
    "state" : "opened",
    "merge_status" : "unchecked",
    "target_project_id" : 143,
    "iid" : 4,
    "description" : "",
    "position" : 0,
    "locked_at" : null,
    "updated_by_id" : null,
    "merge_error" : null,
    "merge_params" : {
      "force_remove_source_branch" : null
    },
    "merge_when_pipeline_succeeds" : false,
    "merge_user_id" : null,
    "merge_commit_sha" : null,
    "deleted_at" : null,
    "in_progress_merge_commit_sha" : null,
    "lock_version" : null,
    "time_estimate" : 0,
    "last_edited_at" : null,
    "last_edited_by_id" : null,
    "head_pipeline_id" : null,
    "source" : {
      "name" : "GitLabPluginTest",
      "description" : "",
      "web_url" : "https://MyGitLab/MyUserName/GitLabPluginTest",
      "avatar_url" : null,
      "git_ssh_url" : "git@MyGitLab/MyUserName/GitLabPluginTest.git",
      "git_http_url" : "https://MyGitLab/MyUserName/GitLabPluginTest.git",
      "namespace" : "MyUserName",
      "visibility_level" : 0,
      "path_with_namespace" : "MyUserName/GitLabPluginTest",
      "default_branch" : "development",
      "homepage" : "https://MyGitLab/MyUserName/GitLabPluginTest",
      "url" : "git@MyGitLab/MyUserName/GitLabPluginTest.git",
      "ssh_url" : "git@MyGitLab/MyUserName/GitLabPluginTest.git",
      "http_url" : "https://MyGitLab/MyUserName/GitLabPluginTest.git"
    },
    "target" : {
      "name" : "GitLabPluginTest",
      "description" : "",
      "web_url" : "https://MyGitLab/MyUserName/GitLabPluginTest",
      "avatar_url" : null,
      "git_ssh_url" : "git@MyGitLab/MyUserName/GitLabPluginTest.git",
      "git_http_url" : "https://MyGitLab/MyUserName/GitLabPluginTest.git",
      "namespace" : "MyUserName",
      "visibility_level" : 0,
      "path_with_namespace" : "MyUserName/GitLabPluginTest",
      "default_branch" : "development",
      "homepage" : "https://MyGitLab/MyUserName/GitLabPluginTest",
      "url" : "git@MyGitLab/MyUserName/GitLabPluginTest.git",
      "ssh_url" : "git@MyGitLab/MyUserName/GitLabPluginTest.git",
      "http_url" : "https://MyGitLab/MyUserName/GitLabPluginTest.git"
    },
    "last_commit" : {
      "id" : "16c979f4a06715d3a3fdd1a40e25ff74a642f79e",
      "message" : "Update .TEST",
      "timestamp" : "2017-07-13T00:02:11+00:00",
      "url" : "https://MyGitLab/MyUserName/GitLabPluginTest/commit/16c979f4a06715d3a3fdd1a40e25ff74a642f79e",
      "author" : {
        "name" : "MyName",
        "email" : "MyEmail"
      }
    },
    "work_in_progress" : false,
    "total_time_spent" : 0,
    "human_total_time_spent" : null,
    "human_time_estimate" : null,
    "url" : "https://MyGitLab/MyUserName/GitLabPluginTest/merge_requests/4",
    "action" : "open"
  },
  "labels" : [ {
    "id" : 380,
    "title" : "release",
    "color" : "#69D100",
    "project_id" : 143,
    "created_at" : "2017-07-11 19:33:50 +0300",
    "updated_at" : "2017-07-11 19:33:50 +0300",
    "template" : false,
    "description" : "",
    "type" : "ProjectLabel",
    "group_id" : null
  } ],
  "repository" : {
    "name" : "GitLabPluginTest",
    "url" : "git@MyGitLab/MyUserName/GitLabPluginTest.git",
    "description" : "",
    "homepage" : "https://MyGitLab/MyUserName/GitLabPluginTest"
  }
}
```

**Comment by omehegan** [comment]
@Zava2012 I'm confused, you mention the GitLab Pull Request Builder plugin. That's different from this GitLab plugin. 

Can you try the latest snapshot that I posted to issue #583?

**Comment by Zava2012** [comment]
I will correct you, **Gitlab Merge Requests Builder** plugin.
True, this is another plugin, but it uses the same variable names. But it does not matter, with the release of the new version 1.4.7, the NPE error has been fixed.

Unfortunately, the problem with the **Filter merge request by label** was not solved by the new version of the plugin. There are no mistakes or anything that could help in solving this problem :(
Without option **Filter merge request by label** everything is still works fine.

**Comment by omehegan** [comment]
We'll have to look into this more, maybe the label functionality is broken.

**Comment by jakub-bochenski** [comment]
@omehegan it is broken: https://github.com/jenkinsci/gitlab-plugin/issues/647

I've found that out while trying to implement #646

**Comment by FabianWilms** [comment]
You have to make sure that the user, which credentials you added under SCM -> Git -> Repositories is actually a member of the project with read access. Even though I added the SSH-Key as a deploy key in Gitlab and Jenkins was able to pull the sources I got the same 404-Error when specifying a branch. After adding the User as a developer to the project the branch selection works as expected.

**Comment by ghost** [comment]
@FabianWilms nailed this issue with his answer.   In gilab the user name in the credential used for git needs to be a member of project in gitlab.   Simply changing this user id resolved the issue for me.

**Comment by omehegan** [comment]
Fixed by #732.
